pkg/handler/collector/s3/s3.go

```diff
@@ -0,0 +1,144 @@
+package s3
+
+import (
+	"context"
+	"fmt"
+	"github.com/guacsec/guac/pkg/handler/collector/s3/download"
+	"github.com/guacsec/guac/pkg/handler/collector/s3/messaging"
+	"github.com/guacsec/guac/pkg/handler/processor"
+	"os"
+	"strings"
+	"sync"
+)
+
+const (
+	S3CollectorType = "S3CollectorType"
+)
+
+type S3Collector struct {
+	config S3CollectorConfig
+}
+
+type S3CollectorConfig struct {
+	MessageProvider     string
+	MessageProviderHost string
+	MessageProviderPort string
+	S3Host              string
+	S3Port              string
+	Region              string // optional (defaults to us-east-1, assumes same region for s3 and sqs)
+	Queues              string
+	MpBuilder           messaging.MessageProviderBuilder // optional
+	DownloaderBuilder   download.BuildDownloader         // optional
+	SigChan             chan os.Signal                   // optional
+}
+
+func NewS3Collector(cfg S3CollectorConfig) (*S3Collector, error) {
+	s3collector := &S3Collector{
+		config: cfg,
+	}
+	return s3collector, nil
+}
+
+func (s S3Collector) RetrieveArtifacts(ctx context.Context, docChannel chan<- *processor.Document) error {
+	queues := strings.Split(s.config.Queues, ",")
+	downloader := getDownloader(s)
+	sigChan := s.config.SigChan
+
+	var wg sync.WaitGroup
+	for _, queue := range queues {
+		wg.Add(1)
+
+		go func(queue string) {
+			mp, err := getMessageProvider(s, queue)
+			if err != nil {
+				fmt.Printf("error getting message provider for queue %v: %v", queue, err)
+				return
+			}
+
+			for {
+				select {
+				case <-sigChan:
+					fmt.Printf("Shutting down collector for queue %s...\n", queue)
+					break
+				default:
+					m, err := mp.ReceiveMessage()
```

Comment:
This is all good. For the future improvement we can consider using one client/connection with multiple queues. It's a better practice in general. But that can be done later.

Reply:
That makes sense :+1: